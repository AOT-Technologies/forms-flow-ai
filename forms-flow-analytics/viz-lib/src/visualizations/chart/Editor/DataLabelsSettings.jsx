import { includes } from "lodash";
import React from "react";
import { useDebouncedCallback } from "use-debounce";
import { Section, Input, Checkbox, ContextHelp } from "@/components/visualizations/editor";
import { EditorPropTypes } from "@/visualizations/prop-types";

export default function DataLabelsSettings({ options, onOptionsChange }) {
  const isShowDataLabelsAvailable = includes(
    ["line", "area", "column", "scatter", "pie", "heatmap"],
    options.globalSeriesType
  );

  const [debouncedOnOptionsChange] = useDebouncedCallback(onOptionsChange, 200);

  return (
    <React.Fragment>
      {isShowDataLabelsAvailable && (
        <Section>
          <Checkbox
            data-test="Chart.DataLabels.ShowDataLabels"
            defaultChecked={options.showDataLabels}
            onChange={event => onOptionsChange({ showDataLabels: event.target.checked })}>
            Show Data Labels
          </Checkbox>
        </Section>
      )}

      <Section>
        <Input
          label={
            <React.Fragment>
              Number Values Format
              <ContextHelp.NumberFormatSpecs />
            </React.Fragment>
          }
          data-test="Chart.DataLabels.NumberFormat"
          defaultValue={options.numberFormat}
          onChange={e => debouncedOnOptionsChange({ numberFormat: e.target.value })}
        />
      </Section>

      <Section>
        <Input
          label={
            <React.Fragment>
              Percent Values Format
              <ContextHelp.NumberFormatSpecs />
            </React.Fragment>
          }
          data-test="Chart.DataLabels.PercentFormat"
          defaultValue={options.percentFormat}
          onChange={e => debouncedOnOptionsChange({ percentFormat: e.target.value })}
        />
      </Section>

      <Section>
        <Input
          label={
            <React.Fragment>
              Date/Time Values Format
              <ContextHelp.DateTimeFormatSpecs />
            </React.Fragment>
          }
          data-test="Chart.DataLabels.DateTimeFormat"
          defaultValue={options.dateTimeFormat}
          onChange={e => debouncedOnOptionsChange({ dateTimeFormat: e.target.value })}
        />
      </Section>

      <Section>
        <Input
          label={
            <React.Fragment>
              Data Labels
              <ContextHelp placement="topRight" arrowPointAtCenter>
                <div style={{ paddingBottom: 5 }}>Use special names to access additional properties:</div>
                <div>
                  <code>{"{{ @@name }}"}</code> series name;
                </div>
                <div>
                  <code>{"{{ @@x }}"}</code> x-value;
                </div>
                <div>
                  <code>{"{{ @@y }}"}</code> y-value;
                </div>
                <div>
                  <code>{"{{ @@yPercent }}"}</code> relative y-value;
                </div>
                <div>
                  <code>{"{{ @@yError }}"}</code> y deviation;
                </div>
                <div>
                  <code>{"{{ @@size }}"}</code> bubble size;
                </div>
                <div style={{ paddingTop: 5 }}>
                  Also, all query result columns can be referenced
                  <br />
                  using
                  <code style={{ whiteSpace: "nowrap" }}>{"{{ column_name }}"}</code> syntax.
                </div>
              </ContextHelp>
            </React.Fragment>
          }
          data-test="Chart.DataLabels.TextFormat"
          placeholder="(auto)"
          defaultValue={options.textFormat}
          onChange={e => debouncedOnOptionsChange({ textFormat: e.target.value })}
        />
      </Section>
    </React.Fragment>
  );
}

DataLabelsSettings.propTypes = EditorPropTypes;
