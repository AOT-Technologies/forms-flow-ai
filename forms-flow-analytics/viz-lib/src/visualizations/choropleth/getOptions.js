import { merge } from "lodash";
import ColorPalette from "./ColorPalette";

const DEFAULT_OPTIONS = {
  mapType: "countries",
  countryCodeColumn: "",
  countryCodeType: "iso_a3",
  valueColumn: "",
  clusteringMode: "e",
  steps: 5,
  valueFormat: "0,0.00",
  noValuePlaceholder: "N/A",
  colors: {
    min: ColorPalette["Light Blue"],
    max: ColorPalette["Dark Blue"],
    background: ColorPalette.White,
    borders: ColorPalette.White,
    noValue: ColorPalette["Light Gray"],
  },
  legend: {
    visible: true,
    position: "bottom-left",
    alignText: "right",
  },
  tooltip: {
    enabled: true,
    template: "<b>{{ @@name }}</b>: {{ @@value }}",
  },
  popup: {
    enabled: true,
    template: "Country: <b>{{ @@name_long }} ({{ @@iso_a2 }})</b>\n<br>\nValue: <b>{{ @@value }}</b>",
  },
};

export default function getOptions(options) {
  return merge({}, DEFAULT_OPTIONS, options);
}
