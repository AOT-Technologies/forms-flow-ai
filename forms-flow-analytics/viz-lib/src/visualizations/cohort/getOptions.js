import { merge } from "lodash";
import ColorPalette from "@/visualizations/ColorPalette";

const DEFAULT_OPTIONS = {
  timeInterval: "daily",
  mode: "diagonal",
  dateColumn: "date",
  stageColumn: "day_number",
  totalColumn: "total",
  valueColumn: "value",

  showTooltips: true,
  percentValues: true,

  timeColumnTitle: "Time",
  peopleColumnTitle: "Users",
  stageColumnTitle: "{{ @ }}",

  numberFormat: "0,0[.]00",
  percentFormat: "0.00%",
  noValuePlaceholder: "-",

  colors: {
    min: "#ffffff",
    max: ColorPalette["Dark Blue"],
    steps: 7,
  },
};

export default function getOptions(options) {
  return merge({}, DEFAULT_OPTIONS, options);
}
