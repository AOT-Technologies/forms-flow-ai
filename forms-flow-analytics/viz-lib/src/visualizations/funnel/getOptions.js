import { isFinite, map, merge, includes } from "lodash";

const DEFAULT_OPTIONS = {
  stepCol: { colName: null, displayAs: "Steps" },
  valueCol: { colName: null, displayAs: "Value" },
  autoSort: true,
  sortKeyCol: { colName: null, reverse: false },
  itemsLimit: 100,
  percentValuesRange: { min: 0.01, max: 1000.0 },
  numberFormat: "0,0[.]00",
  percentFormat: "0[.]00%",
};

export default function getOptions(options, { columns }) {
  options = merge({}, DEFAULT_OPTIONS, options);

  // Validate
  const availableColumns = map(columns, c => c.name);
  if (!includes(availableColumns, options.stepCol.colName)) {
    options.stepCol.colName = null;
  }
  if (!includes(availableColumns, options.valueCol.colName)) {
    options.valueCol.colName = null;
  }
  if (!includes(availableColumns, options.sortKeyCol.colName)) {
    options.sortKeyCol.colName = null;
  }

  if (!isFinite(options.itemsLimit)) {
    options.itemsLimit = DEFAULT_OPTIONS.itemsLimit;
  }
  if (options.itemsLimit < 2) {
    options.itemsLimit = 2;
  }

  if (options.autoSort) {
    options.sortKeyCol.colName = options.valueCol.colName;
    options.sortKeyCol.reverse = true;
  }

  return options;
}
