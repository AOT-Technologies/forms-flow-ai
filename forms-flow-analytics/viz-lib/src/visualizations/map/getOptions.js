import { merge } from "lodash";

const DEFAULT_OPTIONS = {
  latColName: "lat",
  lonColName: "lon",
  classify: null,
  groups: {},
  mapTileUrl: "//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
  clusterMarkers: true,
  customizeMarkers: false,
  iconShape: "marker",
  iconFont: "circle",
  foregroundColor: "#ffffff",
  backgroundColor: "#356AFF",
  borderColor: "#356AFF",
  bounds: null,
  tooltip: {
    enabled: false,
    template: "",
  },
  popup: {
    enabled: true,
    template: "",
  },
};

export default function getOptions(options) {
  options = merge({}, DEFAULT_OPTIONS, options);
  options.mapTileUrl = options.mapTileUrl || DEFAULT_OPTIONS.mapTileUrl;

  // Backward compatibility
  if (options.classify === "none") {
    options.classify = null;
  }

  return options;
}
