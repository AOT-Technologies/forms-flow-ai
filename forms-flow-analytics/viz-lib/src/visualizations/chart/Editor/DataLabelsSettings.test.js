import React from "react";
import enzyme from "enzyme";

import getOptions from "../getOptions";
import DataLabelsSettings from "./DataLabelsSettings";

function findByTestID(wrapper, testId) {
  return wrapper.find(`[data-test="${testId}"]`);
}

function mount(options, done) {
  options = getOptions(options);
  return enzyme.mount(
    <DataLabelsSettings
      visualizationName="Test"
      data={{ columns: [], rows: [] }}
      options={options}
      onOptionsChange={changedOptions => {
        expect(changedOptions).toMatchSnapshot();
        done();
      }}
    />
  );
}

describe("Visualizations -> Chart -> Editor -> Data Labels Settings", () => {
  test("Sets Show Data Labels option", done => {
    const el = mount(
      {
        globalSeriesType: "column",
        showDataLabels: false,
      },
      done
    );

    findByTestID(el, "Chart.DataLabels.ShowDataLabels")
      .last()
      .find("input")
      .simulate("change", { target: { checked: true } });
  });

  test("Changes number format", done => {
    const el = mount(
      {
        globalSeriesType: "column",
        numberFormat: "0[.]0000",
      },
      done
    );

    findByTestID(el, "Chart.DataLabels.NumberFormat")
      .last()
      .simulate("change", { target: { value: "0.00" } });
  });

  test("Changes percent values format", done => {
    const el = mount(
      {
        globalSeriesType: "column",
        percentFormat: "0[.]00%",
      },
      done
    );

    findByTestID(el, "Chart.DataLabels.PercentFormat")
      .last()
      .simulate("change", { target: { value: "0.0%" } });
  });

  test("Changes date/time format", done => {
    const el = mount(
      {
        globalSeriesType: "column",
        dateTimeFormat: "YYYY-MM-DD HH:mm:ss",
      },
      done
    );

    findByTestID(el, "Chart.DataLabels.DateTimeFormat")
      .last()
      .simulate("change", { target: { value: "YYYY MMM DD" } });
  });

  test("Changes data labels format", done => {
    const el = mount(
      {
        globalSeriesType: "column",
        textFormat: null,
      },
      done
    );

    findByTestID(el, "Chart.DataLabels.TextFormat")
      .last()
      .simulate("change", { target: { value: "{{ @@x }} :: {{ @@y }} / {{ @@yPercent }}" } });
  });
});
