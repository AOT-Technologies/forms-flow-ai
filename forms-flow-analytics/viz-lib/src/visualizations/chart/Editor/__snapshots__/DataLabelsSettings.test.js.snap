// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualizations -> Chart -> Editor -> Data Labels Settings Changes data labels format 1`] = `
Object {
  "textFormat": "{{ @@x }} :: {{ @@y }} / {{ @@yPercent }}",
}
`;

exports[`Visualizations -> Chart -> Editor -> Data Labels Settings Changes date/time format 1`] = `
Object {
  "dateTimeFormat": "YYYY MMM DD",
}
`;

exports[`Visualizations -> Chart -> Editor -> Data Labels Settings Changes number format 1`] = `
Object {
  "numberFormat": "0.00",
}
`;

exports[`Visualizations -> Chart -> Editor -> Data Labels Settings Changes percent values format 1`] = `
Object {
  "percentFormat": "0.0%",
}
`;

exports[`Visualizations -> Chart -> Editor -> Data Labels Settings Sets Show Data Labels option 1`] = `
Object {
  "showDataLabels": true,
}
`;
